File: 02_Stack/06_Largest_Rectangle_In_Histogram/84_Largest_Rectangle_In_Histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        s = [] #(index, height)
        max_area = 0

        for i, h in enumerate(heights):
            start = i
            print(i)
            while s and s[-1][1] > h:
                index, height = s.pop()
                max_area = max(max_area, height * (i - index))
                start = index
            s.append((start, h ))
        
        for i, h in s:
            max_area = max(max_area, h * (len(heights) - i))
        
        
        return max_area

File: 02_Stack/06_Largest_Rectangle_In_Histogram/84_Largest_Rectangle_In_Histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<int[]> s = new Stack<>();

        int max_area = 0;

        for (int i = 0; i < heights.length; i++ ){
            int start = i;

            while (!s.isEmpty() && s.peek()[1] > heights[i]){
                int[] top = s.pop();
                int index = top[0];
                int height = top[1];

                max_area = Math.max(max_area, height * (i - index));
                start = index;
            }

            s.add(new int[] {start, heights[i]});
        }

        while (!s.isEmpty()){
            int[] top = s.pop();
            int index = top[0];
            int height = top[1];

            max_area = Math.max(max_area, height * (heights.length - index));
        }

        return max_area;
    }
}
